File: 0053/dp.cpp
#include <cassert>
#include <iostream>
#include <limits>
#include <numeric>
#include <vector>

using namespace std;

class Solution {
 public:
  int maxSubArray(vector<int>& nums) {
    if (nums.size() == 0) return 0;
    // dp[i] means maximum sum of subarray
    vector<int> dp(nums.size());
    dp[0] = nums[0];
    int max_sum = dp[0];
    for (int i = 1; i < nums.size(); i++) {
      // when dp[i - 1] < 0, subarray starts from i
      int sum_so_far = dp[i - 1] > 0 ? dp[i - 1] : 0;
      dp[i] = nums[i] + sum_so_far;
      if (dp[i] > max_sum) max_sum = dp[i];
    }
    return max_sum;
  }
};

int main() {
  Solution sol;
  vector<int> d;

  d = {};
  assert(sol.maxSubArray(d) == 0);

  d = {-2};
  assert(sol.maxSubArray(d) == -2);

  d = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
  assert(sol.maxSubArray(d) == 6);
}

File: 0053/dp.rs
struct Solution;

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        if nums.len() == 0 { return 0; }
        let mut dp: Vec<i32> = vec![0; nums.len()];
        dp[0] = nums[0];
        let mut max: i32 = nums[0];
        for idx in 1..nums.len() {
            if dp[idx-1] > 0 { dp[idx] = dp[idx-1]; }
            dp[idx] += nums[idx];
            max = std::cmp::max(max, dp[idx]);
        }
        return max;
    }
}

fn main() {
    let mut nums;

    nums = vec![-2,1,-3,4,-1,2,1,-5,4];
    assert!(Solution::max_sub_array(nums) == 6);

    nums = vec![-2,1,0,-1,2];
    assert!(Solution::max_sub_array(nums) == 2);

    nums = vec![];
    assert!(Solution::max_sub_array(nums) == 0);
}
